File: 1. Creational Design Patterns/Builder/Python/BuilderTemplate.py
# Builder Pattern – Python Template

class Product:
    def __init__(self):
        self.parts = []
    
    def add_part(self, part):
        self.parts.append(part)
    
    def show(self):
        return f"Product parts: {', '.join(self.parts)}"

class Builder:
    def __init__(self):
        self.product = Product()
    
    def build_part_a(self):
        self.product.add_part("PartA")
        return self
    
    def build_part_b(self):
        self.product.add_part("PartB")
        return self
    
    def build_part_c(self):
        self.product.add_part("PartC")
        return self
    
    def get_result(self):
        return self.product

class Director:
    def __init__(self, builder):
        self.builder = builder
    
    def construct(self):
        return (self.builder
                .build_part_a()
                .build_part_b()
                .build_part_c()
                .get_result())

def main():
    builder = Builder()
    director = Director(builder)
    product = director.construct()
    print(product.show())

if __name__ == "__main__":
    main()

File: 1. Creational Design Patterns/Builder/Kotlin/BuilderTemplate.kt
// Builder Pattern – Kotlin Template
class Product
interface Builder {
    fun buildPartA()
    fun buildPartB()
    fun getResult(): Product
}
class ConcreteBuilder : Builder {
    private val product = Product()
    override fun buildPartA() {}
    override fun buildPartB() {}
    override fun getResult() = product
}
class Director {
    fun construct(b: Builder) {
        b.buildPartA(); b.buildPartB()
    }
}
fun main(){
    val builder: Builder = ConcreteBuilder(); Director().construct(builder)
}
